src/main/java/io/cdap/plugin/http/source/common/DelimitedSchemaDetector.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright © 2024 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.http.source.common;
+
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.FailureCollector;
+import io.cdap.plugin.format.delimited.common.DataTypeDetectorStatusKeeper;
+import io.cdap.plugin.format.delimited.common.DataTypeDetectorUtils;
+import io.cdap.plugin.http.source.batch.HttpBatchSourceConfig;
+
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * Class that detects the schema of the delimited file.
+ */
+public class DelimitedSchemaDetector {
+  public static Schema detectSchema(HttpBatchSourceConfig config, String delimiter, FailureCollector failureCollector) {
+    HttpPartialContentRequest partialContentRequest = new HttpPartialContentRequest(config);
+    HttpPartialContentReader partialContentReader = new HttpPartialContentReader(partialContentRequest);
```

Comment:
Is there no existing class that we can re-use for reading the http content? It seems odd to have a entirely new class with different logic that may potentially read things differently than the reader for the actual data.

Reply:
The problem with the current HTTP reader class is that it performs a basic `GET` request and reads the entire file.

This approach becomes highly inefficient, especially for larger files (CSV/TSV), as it unnecessarily consumes bandwidth. To address this issue, I have developed a new class that allows for partial file reading. This class abstracts the process of reading files partially and provides a getLine() method, which continuously returns the next line of the file.

Reply:
I don't think this partial one works properly though. For example, it doesn't look like it has implemented any of the logic around oauth. I also don't understand how it would be more efficient, as it looks like it is creating a new HttpClient and making a new request for each chunk.

I took a look at the existing logic, it doesn't look like it is doing anything terribly inefficient. RecordPerLinePage is opening an InputStream from the HTTP response and reading it line by line. We should be able to read N lines and close the reader without issue.

Reply:
Why do you suggest this does not work with OAuth 🤔 
I am still using the wrapper to HTTP provided, that abstracts all the proxy and auth logic.

> We should be able to read N lines and close the reader without issue.

Yes this is possible, the class `RecordPerLinePage` does not directly have a methord to return raw lines (which we require) but a new class can be created that extends `RecordPerLinePage`.

One thing also discussed was user accidentally using a wrong or corrupted file with no `\n`.
The buffer-reader will try to read complete file but here we have implement a 10 consumptive chunk loading limit to prevent such issues.

Other that that there is no specific reason to use this class.
What's your suggestion ?

Reply:
I see, thought HttpClient was a library, but it's part of this project, so ignore that comment about OAuth.

I still think its better to re-use the existing classes as much as possible to read line by line. This ensures that the sampling logic is as similar as possible to the actual reading logic. For example, what if the server behind the url doesn't honor the content range headers? Or what if the number of samples is set to a number higher than the results returned in a single page? In that case, it seems like we would want to make requests for additional pages.

I don't think we need to worry too much about the no newline edge case. Setting a 10 chunk limit while sampling but not while reading again makes the behavior inconsistent. If it is really a concern it should be handled in a separate PR, perhaps as an advanced configuration property that applies both to sampling and reading.

Reply:
I have created a new class `RawStringPerLine`, that is a iterator of String, and takes HTTP response as an input.`